### PR TITLE
Replace `gsutil` and `bq` with native Rust

### PR DESCRIPTION
**Pull request by emk**
Whoo, my poor brain. This is a big patch from this week:

> 34 changed files, with 1,771 additions and 519 deletions

This patch gets rid of our dependencies on `bq` and `gsutil`, though our streaming upload and download code is probably slower and less reliable than `gsutil`'s (though it uses far less memory). See the commit message for details. **This is not the final user-ready version.** This is just the first version that passes integration tests.

There's also an unrelated commit that switches us to use `tokio`'s pool-based `spawn_blocking` instead of spawning new threads each time. I thought that we would need that for this patch, because we were going to use a very heavyweight (and seriously out-of-date) synchronous Google Cloud client library, but I decided to just write everything we needed from scratch.

### CODE REVIEW

File: dbcrossbarlib/src/csv_stream.rs [hunk 1]
```diff
@@ -62,45 +61,21 @@ impl CsvStream {
...
-        let forwarder: BoxFuture<()> = async move {
-            try_forward_to_sender(&forwarder_ctx, to_forward, &mut sender).await
-        }
-        .boxed();
-        ctx.spawn_worker(forwarder);
-
-        let stream = receiver.map_err(|err| err.compat());
-        (self.name, Box::pin(stream))
+    ) -> (String, IdiomaticBytesStream) {
+        (self.name, idoimatic_bytes_stream(ctx, self.data))
```
Comment (seamusabshere) [inline dbcrossbarlib/src/csv_stream.rs:78]: typo

File: dbcrossbarlib/src/clouds/gcloud/client.rs [hunk 2]
```diff
@@ -0,0 +1,318 @@
...
+    }
+
+    /// Make an HTTP GET request with the specified URL and query parameters,
+    /// and return the result as a stream.
+    pub(crate) async fn get_response<U, Query>(
+        &self,
+        ctx: &Context,
+        url: U,
+        query: Query,
+    ) -> Result<reqwest::Response>
```
Comment (seamusabshere) [inline dbcrossbarlib/src/clouds/gcloud/client.rs:111]: a little confusing

get_response -> reqwest::Response

post -> Response

File: dbcrossbarlib/src/clouds/gcloud/auth.rs [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+        .ok_or_else(|| {
+            format_err!("cannot find directory to store authentication keys")
+        })?
+        .join("dbcrossbar-gcloud"))
+}
+
+/// Get the service account key needed to connect a server app to BigQuery.
+fn service_account_key() -> Result<ServiceAccountKey> {
+    let json = env::var("GCLOUD_SERVICE_ACCOUNT_KEY")
+        .context("could not find GCLOUD_SERVICE_ACCOUNT_KEY value")?;
```
Comment (seamusabshere) [inline dbcrossbarlib/src/clouds/gcloud/auth.rs:34]: new requirement?

Comment (emk) [reply dbcrossbarlib/src/clouds/gcloud/auth.rs:34]: Yeah, `gcloud auth` uses a whole weird custom auth system that has nothing to do with anybody else, that requires boto config files, lots of env vars, etc. We're going to use very simple env-based auth instead.

File: dbcrossbarlib/src/csv_stream.rs (hunk 1 above)
Comment (emk) [reply dbcrossbarlib/src/csv_stream.rs:78]: Good catch! And once I typoed it, autocomplete kept it consistent everywhere. :-)

**Pull request merged**